src/lib/logical_query_plan/lqp_utils.cpp

```diff
@@ -506,4 +526,32 @@ void remove_invalid_fds(const std::shared_ptr<const AbstractLQPNode>& lqp, std::
    */
 }
 
+std::shared_ptr<AbstractLQPNode> find_diamond_origin_node(
+    const std::shared_ptr<AbstractLQPNode>& union_root_node) {
+  Assert(union_root_node->type == LQPNodeType::Union, "Expecting UnionNode as the diamond's root node.");
+  DebugAssert(union_root_node->input_count() > 1, "Diamond root node does not have two inputs.");
+  bool is_diamond = true;
+  std::optional<std::shared_ptr<AbstractLQPNode>> diamond_origin_node;
+  visit_lqp(union_root_node, [&](const auto& diamond_node) {
+    if (!is_diamond) return LQPVisitation::DoNotVisitInputs;
+    if (diamond_node->output_count() > 1) {
+      if (!diamond_origin_node.has_value()) {
+        diamond_origin_node = diamond_node;
+      } else if (diamond_origin_node != diamond_node) {
+        // The LQP traversal should always end in the same origin node having multiple outputs. Since we found two
+        // different origin nodes, we abort the traversal.
+        is_diamond = false;
+      }
+      return LQPVisitation::DoNotVisitInputs;
+    } else if (!diamond_node->left_input()) {
```

Comment:
This addresses static/stored table nodes in real LQPs, right? Maybe make this explicit by using the type as criterion or add a comment

Reply:
Also MockNodes.

Reply:
Changed to 

```cpp
} else if (!diamond_node->left_input()) {
      // The traversal ends because we reached a MockNode, StoredTableNode or StaticTableNode. Since we did not find a
      // node with multiple outputs yet, union_root_node cannot be considered the root of a diamond.
      is_diamond = false;
    }
```